fdb/fdb.go

```diff
@@ -0,0 +1,236 @@
+package fdb
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/apple/foundationdb/bindings/go/src/fdb"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/directory"
+	"github.com/apple/foundationdb/bindings/go/src/fdb/tuple"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/ipni/dhstore"
+	"github.com/multiformats/go-multicodec"
+	"github.com/multiformats/go-multihash"
+	"lukechampine.com/blake3"
+)
+
+var (
+	_ dhstore.DHStore = (*FDBDHStore)(nil)
+
+	logger        = logging.Logger("store/fdb")
+	fdbHasherPool sync.Pool
+)
+
+const (
+	fdbMaxValueBytes      = 100_000 // 100 KB
+	fdbBlake3HashLength   = 32
+	fdbMultihashDirectory = "mh"
+	fdbMetadataDirectory  = "md"
+)
+
+type FDBDHStore struct {
+	db fdb.Database
+}
+
+func init() {
+	fdbHasherPool.New = func() any {
+		return blake3.New(fdbBlake3HashLength, nil)
+	}
+}
+
+func NewFDBDHStore(o ...Option) (*FDBDHStore, error) {
+	opts, err := newOptions(o...)
+	if err != nil {
+		return nil, err
+	}
+	if err := fdb.APIVersion(opts.apiVersion); err != nil {
+		return nil, err
+	}
+	db, err := fdb.OpenDatabase(opts.clusterFile)
```

Comment:
Am I right in understanding that we are going to have to do an extra hop on the write path? I.e. the write request would hit `dhstore`, then `dhstore` would write into its local FDB and then the local FDB would route that write to the correct node in the consistent hashing ring? 

If so - do we know whether that impacts write throughput in any way? 

Same question with regards to the reads - is there a way to route a read request to the right node in the cluster straight up?

Reply:
There is no local FDB. I believe the client library does batching internally i.e. does not do a remote call per `Set`. 

I think it's far more effective to get this running in `dev` and let the numbers talk.